File: CodingSamples/Java/Platform/SynchronizationTest/Program2.java
class Program {

	static class JointAccount {
		
		private int balance;

		public int balance() { return balance; }

		public boolean withdraw(int amount) {
			boolean result = false;
			//A thread can enter a synchronized block by locking the monitor of
			//the specified object. Only one thread can lock the monitor of a given
			//object at a time while other threads must wait until this thread
			//unlocks the monitor by exiting the synchronized block.
			synchronized(this) {
				if(balance >= amount) {
					Worker.doWork(amount / 500);
					balance -= amount;
					result = true;
				}
			}
			return result;
		}

		public synchronized void deposit(int amount) {
			Worker.doWork(amount / 500);
			balance += amount;
			//allow a thread waiting on the monitor of this object
			//to resume execution
			this.notify();
		}

		public synchronized void withdrawAfterDeposit(int amount) throws InterruptedException {
			do{
				//temporarily unlock the monitor of this object and wait for
				//some other thread to notify the monitor of this object
				this.wait();
			}while(balance < amount);
			balance -= amount;
		}

	}

	public static void main(String[] args) throws Exception {
		var acc = new JointAccount();
		acc.deposit(5000);
		System.out.println("Joint-account opened for Jack and Jill.");
		System.out.printf("Initial balance = %d%n", acc.balance());
		Thread child = new Thread(() -> {
			System.out.println("Jack is withdrawing 3000 from joint account...");
			if(acc.withdraw(3000) == false)
				System.out.println("Jack's withdraw operation failed!");
		});
		child.start();
		System.out.println("Jill is withdrawing 4000 from joint account...");
		if(acc.withdraw(4000) == false)
			System.out.println("Jill's withdraw operation failed!");
		child.join(); //wait for child to exit
		System.out.printf("Final balance = %d%n", acc.balance());
	}
}

File: CodingSamples/DotNet/Platform/MonitorTest/DemoApp/Program2.cs
var acc = new JointAccount();
acc.Deposit(5000);
Console.WriteLine("Joint account opened for Jack and Jill.");
Console.WriteLine("Initial Balance: {0}", acc.Balance);
Thread child = new Thread(() =>
{
    Console.WriteLine("Jack is withdrawing 3000 from the account...");
    if(acc.Withdraw(3000) == false)
        Console.WriteLine("Jack's withdrawal failed!");
});
child.Start();
Console.WriteLine("Jill is withdrawing 4000 from the account...");
if(acc.Withdraw(4000) == false)
    Console.WriteLine("Jill's withdrawal failed!");
child.Join(); //wait for child to exit
Console.WriteLine("Final Balance = {0}", acc.Balance);

class JointAccount
{
    public int Balance { get; set; }

    public bool Withdraw(int amount)
    {
        bool success = false;
        Monitor.Enter(this); //Requires the calling thread to lock the
        //monitor assigned to the specified object. Only a single thread
        //can lock a given monitor at time while other threads must wait 
        //for this thread to unlock it.
        try
        {
            if(Balance >= amount)
            {
                Worker.DoWork(amount / 500);
                Balance -= amount;
                success = true;
            }
        }
        finally
        {
            //unlock the monitor assigned to the specified object
            Monitor.Exit(this); 
        }
        return success;
    }

    public void Deposit(int amount)
    {
        Worker.DoWork(amount / 500);
        Balance += amount;
    }
}
